### ISSUE TITLE
Column "i" does not behave like a column

### ISSUE THREAD

**Issue by alexpeters1208** [deephaven/deephaven-core#757]
**Description**

In attempting to work with table methods that should be able to utilize "i" as an index column, not all functions work as they should.

**Steps to reproduce**

Execute the following commands in the Deephaven Web IDE:

`from deephaven.TableTools import emptyTable, col, newTable`
`t = newTable(col('A',3,6,9,12,15))`
`list = [1,2,3]`
`t1 = t.where("i in list")`

**Expected results**

I expected this command to subset `t` by the indices given in `list`, as would happen if, instead executing the final command given above, you did this:
`t1 = t.updateView("idx = i").where("idx in list").view("A")`

**Actual results**

Instead of getting `t1` as a subset of `t`, I get:
`Type: <class 'RuntimeError'>`
`Value: java.lang.RuntimeException: Column "i" doesn't exist in this table, available columns: [A]`

**Additional details and attachments**

Interestingly, the following command also fails:
`t1 = t.updateView("i = i").where("i in list").view("A")`
with a different error:
`Type: <class 'RuntimeError'>`
`Value: io.deephaven.db.exceptions.ExpressionException: Failed to get expression for all matched patterns`

**Versions**
 - Deephaven: OSS 0.1.1
 - OS: macOS Big Sur 11.4
 - Browser: Safari 14.1.1
 - Docker: 3.3.3

**Comment by niloc132** [comment]
The variable `i` (and `ii`, `k`) should not behave like a column - it is more of an extra local variable to be considered when formulas are evaluated - but the `in` filter syntax is a special kind of filter that only can apply to a column, and so does not work on these magic variables, or any other locally scoped variables.

I'm pretty sure this is all by design - we might want to improve the errors a bit, but it is worth noting that using these variables can cause surprises too - even in cases where they work like `"i == 1 || i == 2 || ..."` this may seem to be broken in certain cases (for example, reverse sort a table that is growing, and notice that rows are not updated to be filtered out when they no longer match). These surprises are also not considered to be bugs - if they were columns then updates would propagate in this way, but they are not.

**Comment by alexpeters1208** [comment]
I see. I guess if they aren't intended to behave like columns, I can't expect any column-like properties outside of the few I'm given. That said, from a user perspective, it is frustrating that what seems to be the most straightforward and obvious way to subset a table by row indices does not work. Maybe there's a smarter, dedicated way to do this, but this is the first thing I reached for and it failed.
Chip asked me to report this as a bug, so I will wait to close this issue until I hear from him. Thanks Colin!

**Comment by niloc132** [comment]
Yeah these behave weirdly, and my general advice would be to avoid them except in very specific circumstances, because they can surprise you. They should be treated like mutable variables that you don't control, with all the caveats that can come from such a thing.

**Comment by rcaudy** [comment]
`in` syntax is a `MatchFilter`, and those don't understand `i`, `ii`, or `k`. I can probably think of a way to "fix" that, but treating them as columns wouldn't be that nice in terms of filter performance. `io.deephaven.db.tables.Table#getSubTable` is a much better way.

**Comment by alexpeters1208** [comment]
Ah, good to know there is a more usable alternative. I had not heard of that method until now. Will look into it.

**Comment by chipkent** [comment]
There are a few things here:
1) i/ii is very useful for historical tables.  Behavior there is well defined.
2) i/ii is garbage for ticking tables.  Also well known.
3) For a user, the behavior of i/ii is closer to a column than a local variable because the value differs by row.
4) Some helper functions around `io.deephaven.db.tables.Table#getSubTable` could reduce the usage of i/ii
5) I would expect `.updateView("i = i")` in the example above to fail, because a column is being set to `i`.
